feat: Priority Donation 구현

hyunmuam opened the pull request:
## 🔗 관련 이슈
<!-- 이슈 번호를 연결해주세요 -->
closes #15 

## 📌 작업 내용
### 구현 사항
<!-- 구현내용에 대해서 설명해주세요. -->
<!-- 예시 :
- 이슈 템플릿, PR 템플릿 작성
- READEME 파일 최초 작성
 --> 
- Priority Donation 구현
  - 락 대기 시 holder에게 우선순위 기부
  - 중첩 기부(Nested Donation) 지원 (최대 depth 8)
  - 다중 기부(Multiple Donation) 지원
- 락 해제 시 우선순위 복원
- `thread_set_priority()` 수정: donation 고려한 우선순위 갱신

### 변경된 파일
<!-- 구현 내용으로 인해 변경된 파일에 대해 설명해주세요 -->
<!-- 예시:
- README.md
  - readme 파일 최초 생성
- issue_template.md
  - 이슈 템플릿 최초 생성
-->
- `threads/thread.c`
  - `thread_set_priority()`: `original_priority` 갱신 후 `refresh_priority()` 호출, `ready_list` 검사 후 필요시 양보
- `threads/thread.h` 
  - struct thread에 priority donation 관련 필드 추가
    - `original_priority`: 기부 전 원본 우선순위
    - `wait_on_lock`: 현재 대기 중인 락
    - `donations`: 이 스레드에게 우선순위를 기부한 스레드들의 리스트
    - `donation_elem`: donations 리스트용 새로운 elem
- `threads/synch.c`
  - 우선순위 기부를 위한 함수 추가
    - `donate_priority()`: 중첩 기부 지원 (최대 depth 8)
    - `refresh_priority()`: donation 리스트 기반 우선순위 갱신
    - `remove_with_lock()`: 락 해제 시 해당 락 관련 donation 제거
  - `lock_acquire()`: 락 대기 시 holder에게 우선순위 기부
  - `lock_release()`: donation 제거 후 우선순위 복원
- `threads/synch.h`
  - 코드 컨벤션에 맞춰 포매팅 적용
### 추가 사항
<!-- 트러블 슈팅 등 추가로 설명할 내용이 있으시다면 아래에 작성해주세요. -->

## ✅ 셀프 체크리스트
<!-- 체크리스트를 모두 완료한 후 제출해주세요 -->
<!-- 완료 처리를 표시하려면 [x], 완료되지 않았다면 [ ] -->
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 주석이나 디버깅 코드를 제거했나요?
- [x] 관련 테스트가 모두 통과하나요?
- [x] 커밋 메시지가 컨벤션에 맞나요?

hyunmuam committed:
style: synch.h 포매팅 규칙 적용

hyunmuam committed:
feat: priority donation 구현

- struct thread에 Priority donation 구현하기 위한 필드 추가 및 초기화
- donate_priority: 우선순위 기부 재귀 함수 (최대 depth 8)
- remove_with_lock: 락 해제 시 해당 락 관련 donation 제거
- refresh_priority: donation 리스트 기반 우선순위 갱신
- lock_acquire/release에 donation 로직 추가
- thread_set_priority: original_priority 갱신 후 선점 검사 추가
- donation_priority_greater: 기부자 우선순위 비교함수 추가

hdh853 commented on threads/synch.c:
```diff
@@ -200,7 +203,17 @@ void lock_acquire(struct lock* lock)
+    struct thread* curr = thread_current();
+    if (lock->holder != NULL) {
+        enum intr_level old = intr_disable();
+        curr->wait_on_lock = lock;
+        list_push_back(&lock->holder->donations, &curr->donation_elem);
+        donate_priority(curr, 0);
+        intr_set_level(old);
+    }
```
여기서 lock->holder 검사를 한 후 인터럽트 비활성화가 되는데, 만약 lock->holder 검사 통과 후 바로 인터럽트가 걸려 조건이 바뀌는 경우가 생기면, 의도하신 동작이 방해받을 수 있다 생각됩니다. 그런 경우는 거의 없겠지만, 그래도 의도하신 방향성엔 '인터럽트를 먼저 걸고 lock->holder의 NULL 검사'가 조금 더 안전하지 않을까? 생각이 듭니다

hyunmuam replied:
@hdh853 
그러네요, `intr_disable`의 범위를 다시 한 번 생각해봐야겠습니다.

hdh853 reviewed (approved):
제가 했던 방식과 구현 로직이 조금 달라도, 함수도 깔끔하게 쪼개주셨고 적절히 주석을 달아주셔서 구현 의도를 이해하기 수월했습니다. 
고생하셨습니다!

hyunmuam committed:
fix: 인터럽트 보호 범위 수정

- lock_acquire(): 인터럽트 보호 범위 변경
- thread_set_priority(): refresh_priority() 호출 시 인터럽트 보호 추가
- lock_try_acquire(): 인터럽트 보호 추가
- lock_release(): 인터럽트 보호 추가
- 공유 자원 접근 시 원자성 보장
